extension/src/util/config.ts

```diff
@@ -26,17 +29,58 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
     return vscode.workspace.getConfiguration("java").get<string | null>("import.gradle.java.home", null);
 }
 
-export function getConfigGradleJavaHome(): string | null {
-    return getConfigJavaImportGradleJavaHome() || getJdtlsConfigJavaHome() || getConfigJavaHome();
+export function getJavaExecutablePathFromJavaHome(javaHome: string): string {
+    return path.join(javaHome, "bin", JAVA_FILENAME);
+}
+
+export async function findValidJavaHome(): Promise<string | undefined> {
+    const javaHomeGetters = [getJdtlsConfigJavaHome, getConfigJavaHome, getConfigJavaImportGradleJavaHome];
+    let javaHome: string | undefined = undefined;
+    let javaVersion = 0;
+
+    for (const getJavaHome of javaHomeGetters) {
+        javaHome = getJavaHome() || undefined;
+        if (javaHome) {
+            javaVersion = await getMajorVersion(javaHome);
+            if (javaVersion >= REQUIRED_JDK_VERSION) {
+                return javaHome;
+            }
+        }
+    }
+
+    // Search valid JDKs from env.JAVA_HOME, env.PATH, SDKMAN, jEnv, jabba, common directories
+    const javaRuntimes = await listJdks();
+    const validJdks = javaRuntimes.filter((r) => r.version!.major >= REQUIRED_JDK_VERSION);
```

Comment:
Use `.find()` to speed up. since you only need the first element meets the requirement